File: Longest ZigZag Path in a Binary Tree/longest-zigzag-path-in-a-binary-tree.py
# Definition for a binary tree node.
class TreeNode(object):
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution(object):
    
    def longestPath(self, r, dir, currlen, maxPath):
        if not r:
            return
        
        maxPath[0] = max(maxPath[0], currlen)
        
        self.longestPath(r.left, 'l', 1 if dir == 'l' else currlen + 1, maxPath)
        
        self.longestPath(r.right, 'r', 1 if dir == 'r' else currlen + 1, maxPath)

    def longestZigZag(self, root):
        maxPath = [0]
        self.longestPath(root, 'l', 0, maxPath)
        self.longestPath(root, 'r', 0, maxPath)
        return maxPath[0]

File: Longest ZigZag Path in a Binary Tree/longest-zigzag-path-in-a-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    
    void longestPath(TreeNode* r, char dir,int currlen, int &maxPath){
        if(r==NULL)
        return;
        maxPath=max(maxPath,currlen);
        longestPath(r->left,'l',dir=='l'?1:currlen+1,maxPath);
        longestPath(r->right,'r',dir=='r'?1:currlen+1,maxPath);
    }
    int longestZigZag(TreeNode* root) {
        int maxPath=0;
        longestPath(root,'l',0, maxPath);
        longestPath(root,'r',0, maxPath);
        return maxPath;
    }
};